src/VisualStudio/Core/Def/Storage/VisualStudioPersistentStorageLocationService.cs

```diff
@@ -5,19 +5,193 @@
 using Microsoft.CodeAnalysis.Host;
 using Microsoft.CodeAnalysis;
 using Microsoft.VisualStudio.LanguageServices.Implementation.ProjectSystem;
+using Microsoft.VisualStudio.LanguageServices.Implementation.Extensions;
+using System;
+using Microsoft.CodeAnalysis.Editor.Shared.Utilities;
+using Microsoft.VisualStudio.Shell;
+using Microsoft.VisualStudio.Shell.Interop;
+using Microsoft.Internal.VisualStudio.Shell.Interop;
 
 namespace Microsoft.VisualStudio.LanguageServices.Storage
 {
     [ExportWorkspaceService(typeof(IPersistentStorageLocationService), ServiceLayer.Host), Shared]
-    internal class VisualStudioPersistentStorageLocationService : IPersistentStorageLocationService
+    internal class VisualStudioPersistentStorageLocationService : ForegroundThreadAffinitizedObject, IPersistentStorageLocationService
     {
-        public bool IsSupported(Microsoft.CodeAnalysis.Workspace workspace)
+        private readonly IServiceProvider _serviceProvider;
+
+        private readonly object _gate = new object();
+        private SolutionId _currentSolutionId = null;
+        private string _currentWorkingFolderPath = null;
+        private bool _solutionEventsAdvised = false;
+
+        public event EventHandler<PersistentStorageLocationChangingEventArgs> StorageLocationChanging;
```

Comment:
can you give a threading comment on this?  like which threads this might fire on.